Index: pkg-zfs/cmd/Makefile.am
===================================================================
--- pkg-zfs.orig/cmd/Makefile.am	2015-03-02 22:29:17.890980774 +0100
+++ pkg-zfs/cmd/Makefile.am	2015-03-02 22:31:29.797322487 +0100
@@ -1,2 +1,3 @@
 SUBDIRS  = zfs zpool zdb zhack zinject zstreamdump ztest zpios
 SUBDIRS += mount_zfs fsck_zfs zvol_id vdev_id arcstat dbufstat zed
+SUBDIRS += arc_summary
Index: pkg-zfs/configure.ac
===================================================================
--- pkg-zfs.orig/configure.ac	2015-03-02 22:30:35.214008692 +0100
+++ pkg-zfs/configure.ac	2015-03-02 22:31:57.248979249 +0100
@@ -110,6 +110,7 @@
 	cmd/vdev_id/Makefile
 	cmd/arcstat/Makefile
 	cmd/dbufstat/Makefile
+	cmd/arc_summary/Makefile
 	cmd/zed/Makefile
 	contrib/Makefile
 	contrib/bash_completion.d/Makefile
